core/arch/aarchxx/mangle.c

```diff
@@ -2922,7 +2954,573 @@ mangle_icache_op(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+        /* We pass false to avoid the status reg, which we ourselves use. */
+        scratch = pick_scratch_reg(dcontext, instr, swap_reg, DR_REG_NULL, DR_REG_NULL,
+                                   false, &slot, &should_restore);
+        insert_save_to_tls_if_necessary(dcontext, ilist, instr, scratch, slot);
+#ifdef ARM
+        if (!is_cbnz_available(dcontext, reg_res)) {
+            /* We have no CBNZ so we need to preserve the flags. */
+            dr_save_arith_flags_to_reg(dcontext, ilist, instr, scratch);
+            PRE(ilist, instr,
+                instr_create_save_to_tls(dcontext, scratch, TLS_LDSTEX_FLAGS_SLOT));
+        }
+#endif
+        if (is_pair) {
+            scratch2 = pick_scratch_reg(dcontext, instr, swap_reg, scratch, DR_REG_NULL,
+                                        false, &slot2, &should_restore2);
+            scratch3 = pick_scratch_reg(dcontext, instr, swap_reg, scratch, scratch2,
+                                        false, &slot3, &should_restore3);
+            insert_save_to_tls_if_necessary(dcontext, ilist, instr, scratch2, slot2);
+            insert_save_to_tls_if_necessary(dcontext, ilist, instr, scratch3, slot3);
+        }
+        /* Compare address, arranging op_res to show failure on mismatch.
+         * XXX i#400: It is possible that the store could fault and the app could
+         * examine ops_res in the handler: i.e., it's not truly dead.  We do not
+         * account for that here.
+         */
+        PRE(ilist, instr,
+            instr_create_restore_from_tls(dcontext, scratch, TLS_LDSTEX_ADDR_SLOT));
+        insert_compare_and_jump_not_equal(dcontext, ilist, instr, op_res,
+                                          opnd_create_reg(reg_base),
+                                          opnd_create_reg(scratch), no_match);
+
+        /* Compare size, arranging op_res to show failure on mismatch. */
+        PRE(ilist, instr,
+            instr_create_restore_from_tls(dcontext, scratch, TLS_LDSTEX_SIZE_SLOT));
+        insert_compare_and_jump_not_equal(
+            dcontext, ilist, instr, op_res, opnd_create_reg(scratch),
+            OPND_CREATE_INT(reg_get_size(reg_st_val)), no_match);
+    } else {
+        /* We aren't saving the flags so we can only handle Thumb mode with CBNZ. */
+        IF_ARM(ASSERT(is_cbnz_available(dcontext, reg_res)));
+    }
+
+    /* Perform the compare-and-swap. */
+    if (!ldex_in_same_block) {
+        PRE(ilist, instr,
+            instr_create_restore_from_tls(dcontext, scratch, TLS_LDSTEX_VALUE_SLOT));
+        reg_orig_ld_val = scratch;
+        if (is_pair) {
+            PRE(ilist, instr,
+                instr_create_restore_from_tls(dcontext, scratch2,
+                                              TLS_LDSTEX_VALUE2_SLOT));
+            reg_orig_ld_val2 = scratch2;
+        }
+    }
+    reg_id_t reg_new_ld_val;
+    reg_id_t reg_new_ld_val2 = is_pair ? scratch3 : DR_REG_NULL;
```

Comment:
`scratch3` is currently reserved only when `!ldex_in_same_block`

Reply:
Nice: this is a bug.  Looking carefully at the test: both pair instances are !same-block.  I changed the test to have one of them be same-block, and added scratch3 for that case.  (Note that it shows up as an assert with scratch3 as DR_REG_NULL so at least in debug build it shows up up front.)